File: packages/hydra-rust/src/hydra/lib/sets.rs
// hydra.lib.sets primitives

use std::collections::BTreeSet;

pub fn contains<A: Ord>(x: A, s: BTreeSet<A>) -> bool {
    s.contains(&x)
}

pub fn difference<A: Ord + Clone>(s1: BTreeSet<A>, s2: BTreeSet<A>) -> BTreeSet<A> {
    s1.difference(&s2).cloned().collect()
}

pub fn empty<A: Ord>() -> BTreeSet<A> {
    BTreeSet::new()
}

pub fn from_list<A: Ord>(l: Vec<A>) -> BTreeSet<A> {
    l.into_iter().collect()
}

pub fn insert<A: Ord>(x: A, s: BTreeSet<A>) -> BTreeSet<A> {
    let mut result = s;
    result.insert(x);
    result
}

pub fn intersection<A: Ord + Clone>(s1: BTreeSet<A>, s2: BTreeSet<A>) -> BTreeSet<A> {
    s1.intersection(&s2).cloned().collect()
}

pub fn is_empty<A>(s: BTreeSet<A>) -> bool {
    s.is_empty()
}

pub fn map<A: Ord, B: Ord>(f: impl Fn(A) -> B, s: BTreeSet<A>) -> BTreeSet<B> {
    s.into_iter().map(f).collect()
}

pub fn member<A: Ord>(x: A, s: BTreeSet<A>) -> bool {
    s.contains(&x)
}

pub fn singleton<A: Ord>(x: A) -> BTreeSet<A> {
    let mut s = BTreeSet::new();
    s.insert(x);
    s
}

pub fn size<A>(s: BTreeSet<A>) -> i32 {
    s.len() as i32
}

pub fn to_list<A: Clone>(s: BTreeSet<A>) -> Vec<A> {
    s.into_iter().collect()
}

pub fn union<A: Ord + Clone>(s1: BTreeSet<A>, s2: BTreeSet<A>) -> BTreeSet<A> {
    s1.union(&s2).cloned().collect()
}

File: packages/hydra-typescript/src/main/typescript/hydra/lib/sets.ts
/**
 * Set primitive functions for Hydra-TypeScript.
 */

export function empty<T>(): ReadonlySet<T> {
  return new Set();
}

export function singleton<T>(x: T): ReadonlySet<T> {
  return new Set([x]);
}

export function fromList<T>(xs: ReadonlyArray<T>): ReadonlySet<T> {
  return new Set(xs);
}

export function toList<T>(s: ReadonlySet<T>): ReadonlyArray<T> {
  return [...s];
}

export function insert<T>(x: T, s: ReadonlySet<T>): ReadonlySet<T> {
  const result = new Set(s);
  result.add(x);
  return result;
}

export function member<T>(x: T, s: ReadonlySet<T>): boolean {
  return s.has(x);
}

export function size<T>(s: ReadonlySet<T>): number {
  return s.size;
}

export function union<T>(
  a: ReadonlySet<T>,
  b: ReadonlySet<T>,
): ReadonlySet<T> {
  const result = new Set(a);
  for (const x of b) {
    result.add(x);
  }
  return result;
}

export function intersection<T>(
  a: ReadonlySet<T>,
  b: ReadonlySet<T>,
): ReadonlySet<T> {
  const result = new Set<T>();
  for (const x of a) {
    if (b.has(x)) {
      result.add(x);
    }
  }
  return result;
}

export function difference<T>(
  a: ReadonlySet<T>,
  b: ReadonlySet<T>,
): ReadonlySet<T> {
  const result = new Set<T>();
  for (const x of a) {
    if (!b.has(x)) {
      result.add(x);
    }
  }
  return result;
}

export function isEmpty<T>(s: ReadonlySet<T>): boolean {
  return s.size === 0;
}
